crates/bevy_ptr/src/lib.rs

```diff
@@ -1326,18 +1324,43 @@ macro_rules! move_as_ptr {
 /// [`assign_to`]: MovingPtr::assign_to
 #[macro_export]
 macro_rules! deconstruct_moving_ptr {
-    ($ptr:ident => {$($field_name:ident,)*}) => {
+    ($ptr:ident => {$($field_name:ident),* $(,)?}) => {
         $crate::deconstruct_moving_ptr!($ptr => ($($field_name => $field_name,)*))
     };
-    ($ptr:ident => ($($field_index:tt => $field_alias:ident,)*)) => {
-        $(let $field_alias = $ptr.move_field(|f| &raw mut (*f).$field_index);)*
-        core::mem::forget($ptr);
+    ($ptr:ident => ($($field_index:tt => $field_alias:ident),* $(,)?)) => {
+        // Specify the type to make sure we don't `mem::forget` a mere `&mut MovingPtr`
+        let mut ptr: $crate::MovingPtr<_, _> = $ptr;
+        let _ = || {
+            let value = &mut *ptr;
+            // If a field is mentioned twice, this will cause mutable aliasing and fail to compile.
+            let _ = ($(&mut value.$field_index),*);
+        };
+        // SAFETY:
+        // - `f` does a raw pointer offset, which always returns a non-null pointer to a field inside `T`
+        // - The struct is not `repr(packed)`, since otherwise the block of code above would fail compilation
+        // - We call `mem::forget` on `self` immediately after these calls
+        // - Each field is distinct, since otherwise the block of code above would fail compilation
+        $(let $field_alias = unsafe { ptr.move_field(|f| &raw mut (*f).$field_index) };)*
```

Comment:
I believe this is most likely unsound. Doing `&raw mut (*f).$field_index` does not guarantee you'are actually accessing a field on the type pointed by `f`, but instead you might trigger an an autoderef, which will call `DerefMut` with a `&mut` reference to your type. That in turn allows to invalidate all the previous pointers you obtained. This will most likely trigger the (deny-by-default) `dangerous_implicit_autorefs` lint, but lints cannot guarantee soundness.

Destructuring as mentioned in the other comment should solve this issue, even if never executed (e.g. in the above closure), as that still statically forces the idents to be valid fields of your type.

Reply:
> Oh! Yeah, you're right! I would not have expected that the new syntax added for the purpose of creating raw pointers without creating references would... still create references sometimes.

That's because the reference is created before the raw pointer operator is applied. The operator acts on a _place_ and creating a place to `(*f).field` might involve calling `Deref`/`DerefMut` and creating a reference.

I'm not sure where this is mentioned for `&raw`/`&raw mut`, but it's mentioned in the ["Safety" section](https://doc.rust-lang.org/stable/std/ptr/macro.addr_of_mut.html#safety) of `addr_of!`/`addr_of_mut!` (which are the same as the other operators in every aspect).